### PR TITLE
FRDM-K22F Support

### PR DESCRIPTION
**Pull request by teburd**
Depends on https://github.com/zephyrproject-rtos/hal_nxp/pull/10 (merged)

Working through a few more test cases that are failing but all the kernel tests now pass, Draft until all tests pass or have clear explanations for failures

I would really like to get this in for 2.1 and have marked it as such, as I have a product based on this chip.

### CODE REVIEW

**Review by galak — CHANGES_REQUESTED** [review]
some updates to board.yaml should fix things right up.

**Comment by bfrog** [comment]
@galak all fixed up

File: soc/arm/nxp_kinetis/k2x/Kconfig.soc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	select HAS_MCUX
+	select HAS_MCUX_SMC
+	select HAS_MCUX_ADC16
+	select HAS_MCUX_FTFX
+	select HAS_MCUX_FTM
+	select HAS_MCUX_RNGA
+	select HAS_MCUX_SIM
+	select HAS_OSC
+	select HAS_MCG
+    select HAS_CMSIS
```
Comment (MaureenHelm) [inline soc/arm/nxp_kinetis/k2x/Kconfig.soc]: Replace the spaces with a tab

File: soc/arm/nxp_kinetis/k2x/linker.ld [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * They are loaded at reset to various Flash Memory module (FTFE) registers.
+ */
+
+/*
+ * No need to account for this when running a RAM-only image since that
+ * security feature resides in ROM.
+ */
+#if defined(CONFIG_XIP)
+	#define SKIP_TO_KINETIS_FLASH_CONFIG . = 0x400;
+#endif
```
Comment (MaureenHelm) [inline soc/arm/nxp_kinetis/k2x/linker.ld]: Remove this if #19094 gets merged first

File: soc/arm/nxp_kinetis/k2x/soc.c [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+ * The structure is:
+ * -Backdoor Comparison Key for unsecuring the MCU - 8 bytes
+ * -Program flash protection bytes, 4 bytes, written to FPROT0-3
+ * -Flash security byte, 1 byte, written to FSEC
+ * -Flash nonvolatile option byte, 1 byte, written to FOPT
+ * -Reserved, 1 byte, (Data flash protection byte for FlexNVM)
+ * -Reserved, 1 byte, (EEPROM protection byte for FlexNVM)
+ *
+ */
+u8_t __kinetis_flash_config_section __kinetis_flash_config[] = {
```
Comment (MaureenHelm) [inline soc/arm/nxp_kinetis/k2x/soc.c]: Remove this if #19094 gets merged first

File: boards/arm/frdm_k22f/frdm_k22f.dts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+		int1-gpios = <&gpiod 0 0>;
+		int2-gpios = <&gpiod 1 0>;
+	};
+};
+
+&spi0 {
+	status = "okay";
+	cs-gpios = <&gpiod 4 0>;
+
+	mcr20a@0 {
```
Comment (MaureenHelm) [inline boards/arm/frdm_k22f/frdm_k22f.dts]: Remove the mcr20a node. See #17407

File: boards/arm/frdm_k22f/frdm_k22f.yaml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  - gnuarmemb
+supported:
+  - adc
+  - i2c
+  - nvs
+  - spi
+  - gpio
+  - usb_device
+  - watchdog
+testing:
```
[the remark this answers is not in the document]
Comment (MaureenHelm) [reply boards/arm/frdm_k22f/frdm_k22f.yaml]: It needs to be moved out of the `testing` section.
https://github.com/zephyrproject-rtos/zephyr/blob/69ba9dafcbd759bcadc4f8912a419428ef1ed167/boards/arm/mimxrt1050_evk/mimxrt1050_evk.yaml#L7-L16

File: boards/arm/frdm_k22f/frdm_k22f_defconfig [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+CONFIG_SOC_SERIES_KINETIS_K2X=y
+CONFIG_BOARD_FRDM_K22F=y
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+CONFIG_SERIAL=y
+CONFIG_CORTEX_M_SYSTICK=y
+CONFIG_GPIO=y
+CONFIG_PINMUX=y
+CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=72000000
+CONFIG_OSC_LOW_POWER=y
```
Comment (MaureenHelm) [inline boards/arm/frdm_k22f/frdm_k22f_defconfig]: Add `CONFIG_ARM_MPU=y`

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
A few minor nits, but otherwise this looks really great. Thanks for working on it!

File: boards/arm/frdm_k22f/frdm_k22f_defconfig (hunk 6 above)
Comment (bfrog) [reply boards/arm/frdm_k22f/frdm_k22f_defconfig]: So unfortunately mcuxpresso sdk doesn't seem to actually have any of the MPU definitions setup, so when I enable this SYSMPU in particular isn't defined anywhere

Comment (bfrog) [reply boards/arm/frdm_k22f/frdm_k22f_defconfig]: I've filed a ticket to look into this
https://github.com/zephyrproject-rtos/zephyr/issues/19152

File: boards/arm/frdm_k22f/doc/index.rst [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
++-----------+------------+-------------------------------------+
+| FLASH     | on-chip    | soc flash                           |
++-----------+------------+-------------------------------------+
+| USB       | on-chip    | USB device                          |
++-----------+------------+-------------------------------------+
+| SENSOR    | off-chip   | fxos8700 polling;                   |
+|           |            | fxos8700 trigger                    |
++-----------+------------+-------------------------------------+
+
+The default configuration can be found in the defconfig file:
```
Comment (dbkinder) [inline boards/arm/frdm_k22f/doc/index.rst]: Let's change this to link to the defconfig file:
```
The default configuration can be found in the defconfig file:
:zephyr_file:`boards/arm/frdm_k22f/frdm_k22f_defconfig`.
```

**Review by dbkinder — CHANGES_REQUESTED** [review]
one doc tweak, otherwise LGTM

File: boards/arm/frdm_k22f/doc/index.rst (hunk 7 above)
Comment (bfrog) [reply boards/arm/frdm_k22f/doc/index.rst]: Thats a nice little tag, I'll try to remember that, will do and thanks for reviewing

Comment (bfrog) [reply boards/arm/frdm_k22f/doc/index.rst]: Of course, thanks for looking at this

File: soc/arm/nxp_kinetis/k2x/soc.c [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+ * This module provides routines to initialize and support board-level
+ * hardware for the fsl_frdm_k22f platform.
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <init.h>
+#include <soc.h>
+#include <uart.h>
+#include <linker/sections.h>
```
Comment (MaureenHelm) [inline soc/arm/nxp_kinetis/k2x/soc.c]: Remove this

File: soc/arm/nxp_kinetis/k2x/soc.c [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+	CLOCK_InitOsc0(&oscConfig);
+	CLOCK_SetXtal0Freq(CONFIG_OSC_XTAL0_FREQ);
+
+
+	CLOCK_SetInternalRefClkConfig(kMCG_IrclkEnable, kMCG_IrcSlow,
+				      CONFIG_MCG_FCRDIV);
+
+	/* Configure FLL external reference divider (FRDIV). */
+	/*CLOCK_CONFIG_SetFllExtRefDiv(0);*/
+	MCG->C1 = ((MCG->C1 & ~MCG_C1_FRDIV_MASK) | MCG_C1_FRDIV(0x0U));
```
Comment (MaureenHelm) [inline soc/arm/nxp_kinetis/k2x/soc.c]: Why did you comment out the CLOCK_ call and modify the MCG register directly?

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: soc/arm/nxp_kinetis/k2x/soc.c (hunk 9 above)
Comment (bfrog) [reply soc/arm/nxp_kinetis/k2x/soc.c]: This is leftover from the original PR created by prevas I believe, will fix it

**Review by MaureenHelm — APPROVED** [review]
@galak @dbkinder please revisit

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/frdm_k22f/doc/index.rst [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+.. image:: frdm_k22f.jpg
+   :width: 720px
+   :align: center
+   :alt: FRDM-K22F
+
+Hardware
+********
+
+- MK22FN512VLH12 (120 MHz, 1 MB flash memory, 256 KB RAM, low-power,
+  crystal-less USB, and 22 Low profile Quad Flat Package (LQFP))
```
[the remark this answers is not in the document]
Comment (bfrog) [reply boards/arm/frdm_k22f/doc/index.rst]: Thanks, this is a find/replace mistake (replacing 64's with 22's), very possibly my own flub, good catch!

**Review by dbkinder — APPROVED** [review]
+1 for docs, thanks!

**Review by galak — APPROVED** [review]

**Pull request merged**